services/headless-lms/server/src/domain/exercise_repositories.rs

```diff
@@ -0,0 +1,420 @@
+use anyhow::Context;
+use blake3::Hash;
+use git2::{build::RepoBuilder, Cred, FetchOptions, RemoteCallbacks, Repository};
+use headless_lms_models::{exercise_repositories, repository_exercises};
+use headless_lms_utils::{
+    file_store::{self, FileStore},
+    folder_checksum, ApplicationConfiguration,
+};
+use sqlx::{Acquire, PgConnection};
+use std::{
+    collections::HashMap,
+    io::Cursor,
+    path::{Path, PathBuf},
+};
+use uuid::Uuid;
+use walkdir::{DirEntry, WalkDir};
+
+pub struct RepositoryExercise {
+    pub url: String,
+}
+
+/// Processes an exercise repository, creating a repository exercise for each exercise in it.
+/// Each exercise is compressed and uploaded to file storage.
+pub async fn process(
+    conn: &mut PgConnection,
+    repository_id: Uuid,
+    url: &str,
+    deploy_key: Option<&str>,
+    file_store: &dyn FileStore,
+    app_conf: &ApplicationConfiguration,
+) -> anyhow::Result<Vec<RepositoryExercise>> {
+    let mut stored_files = vec![];
+    match process_inner(
+        conn,
+        repository_id,
+        url,
+        deploy_key,
+        file_store,
+        &mut stored_files,
+        app_conf,
+    )
+    .await
+    {
+        Ok(res) => {
+            exercise_repositories::mark_success(conn, repository_id).await?;
+            Ok(res)
+        }
+        Err(err) => {
+            if !stored_files.is_empty() {
+                warn!("Failed while creating new exercise repository, cleaning files that were uploaded");
+                for file in stored_files {
+                    if let Err(err) = file_store.delete(&file).await {
+                        error!("Failed to clean file {}: {err}", file.display());
+                    }
+                }
+            }
+            exercise_repositories::mark_failure(conn, repository_id, &err.to_string()).await?;
+            Err(err)
+        }
+    }
+}
+
+// implements the logic for process so that we can conveniently handle all errors in process
+async fn process_inner(
+    conn: &mut PgConnection,
+    repository_id: Uuid,
+    url: &str,
+    deploy_key: Option<&str>,
+    file_store: &dyn FileStore,
+    stored_files: &mut Vec<PathBuf>,
+    app_conf: &ApplicationConfiguration,
+) -> anyhow::Result<Vec<RepositoryExercise>> {
+    let mut tx = conn.begin().await?;
+
+    // clone repo to temp dir
+    let temp = tempfile::tempdir()?;
```

Comment:
One question for the future is do we want to cache the cloned repos simmilar to tmc-server is doing?